tests/apps/qchem/test_dynamics.py

```diff
@@ -94,3 +100,91 @@ def test_evolution(time, unitary, frequency, prob):
         p = eng.run(gbs).state.all_fock_probs()
 
     assert np.allclose(p, prob)
+
+
+@pytest.mark.parametrize("d", [d1, d2])
+class TestSampleFock:
+    """Tests for the function ``strawberryfields.apps.qchem.dynamics.sample_fock``"""
+
+    def test_invalid_n_samples(self, d):
+        """Test if function raises a ``ValueError`` when a number of samples less than one is
+        requested."""
+        with pytest.raises(ValueError, match="Number of samples must be at least one"):
+            in_state, t, U, w, ns = d
+            dynamics.sample_fock(in_state, t, U, w, 0)
+
+    def test_invalid_time(self, d):
+        """Test if function raises a ``ValueError`` when a negative time is given."""
+        with pytest.raises(ValueError, match="Time must be zero or positive"):
+            in_state, t, U, w, ns = d
+            dynamics.sample_fock(in_state, -t, U, w, ns)
+
+    def test_negative_frequency(self, d):
+        """Test if function raises a ``ValueError`` when negative frequencies are given."""
+        with pytest.raises(ValueError, match="Vibrational frequencies must be larger than zero"):
+            in_state, t, U, w, ns = d
+            dynamics.sample_fock(in_state, t, U, -w, ns)
+
+    def test_zero_frequency(self, d):
+        """Test if function raises a ``ValueError`` when zero frequencies are given."""
+        with pytest.raises(ValueError, match="Vibrational frequencies must be larger than zero"):
+            in_state, t, U, w, ns = d
+            dynamics.sample_fock(in_state, t, U, 0.0 * w, ns)
+
+    def test_complex_unitary(self, d):
+        """Test if function raises a ``ValueError`` when a complex unitary is given."""
+        with pytest.raises(
+            ValueError, match="The normal mode to local mode transformation matrix must be real"
+        ):
+            in_state, t, U, w, ns = d
+            dynamics.sample_fock(in_state, t, 1.0j * U, w, ns)
+
+    def test_invalid_loss(self, d):
+        """Test if function raises a ``ValueError`` when the loss parameter is specified outside
+        of range."""
+        with pytest.raises(
+            ValueError, match="Loss parameter must take a value between zero and one"
+        ):
+            in_state, t, U, w, ns = d
+            dynamics.sample_fock(in_state, t, U, w, ns, w[-1])
+
+    def test_loss(self, monkeypatch, d):
+        """Test if function correctly creates the SF program for lossy circuits."""
+
+        def save_hist(*args):
+            call_history.append(args[1])
+            return sf.engine.Result
+
+        call_history = []
+        with monkeypatch.context() as m:
+            m.setattr(sf.engine.Result, "samples", np.array([[0]]))
+            m.setattr(sf.LocalEngine, "run", save_hist)
+            dynamics.sample_fock(*d, loss=0.5)
```

Comment:
I'm a bit confused what these two tests are doing, they seem similar to `test_loss` and `test_no_loss` that test the `sample.sample` function, but there there are asserts to check what the program looks like.

Reply:
I have added the asserts. They were missing by mistake.